File: 4-Sliding Window/1-buy_sell_stock.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

int maxProfit(vector<int>& prices) {
    int left = 0;
    int right = 1;
    int max_prof = 0;
    int len = prices.size();

    while (right < len) {
        if (prices[left] < prices[right]) {
            int profit = prices[right] - prices[left];
            max_prof = max(profit, max_prof);
        } else {
            left = right;
        }
        right++;
    }
    return max_prof;       
}

int main() {
    vector<int> prices = {7,1,5,3,6,4};

    cout << maxProfit(prices);

    return 0;
}

File: PYTHON/4-SlidingWindow/1-buy_sell_stock.py
from typing import List

def maxProfit(prices: List[int]) -> int:
    left, right = 0, 1
    max_prof = 0

    while right < len(prices):
        if prices[left] < prices[right]:
            profit = prices[right] - prices[left]
            max_prof = max(profit, max_prof)
        else:
            left = right
        right += 1

    return max_prof

# dp (dinamic programming)
# def maxProfit(prices: List[int]) -> int:
#     max_prof = 0
#     min_buy = prices[0]

#     for sell in prices:
#         max_prof = max(max_prof, sell - min_buy)
#         min_buy = min(min_buy, sell)
    
#     return max_prof

if __name__ == "__main__":
    prices = [10, 1, 5, 6, 7, 1]
    print(maxProfit(prices))